File: tests/test_palanca_secuencial.py
"""
Tests para palanca_secuencial.ino
Verifican que el código fuente cumple la especificación de hardware:
  - Pin 6 = Subir (botón 0)
  - Pin 3 = Bajar  (botón 1)
  - 2 botones HID
  - Usa INPUT_PULLUP (común a GND)
  - Envía estado solo cuando hay cambio
"""

import re
from pathlib import Path

INO = Path(__file__).parent.parent / "palanca_secuencial" / "palanca_secuencial.ino"


def source() -> str:
    return INO.read_text(encoding="utf-8")


# ── Existencia del archivo ────────────────────────────────────────────────────

def test_archivo_existe():
    assert INO.exists(), f"No se encontró {INO}"


# ── Dependencia HID ───────────────────────────────────────────────────────────

def test_incluye_joystick_library():
    assert '#include <Joystick.h>' in source(), \
        "Debe incluir la ArduinoJoystickLibrary (#include <Joystick.h>)"


# ── Configuración del joystick ────────────────────────────────────────────────

def test_joystick_tiene_2_botones():
    """El tercer argumento del constructor Joystick_ es el número de botones."""
    match = re.search(r'Joystick_\s*\([^)]+\)', source(), re.DOTALL)
    assert match, "No se encontró el constructor Joystick_()"
    args = [a.strip() for a in match.group().split(',')]
    # args[2] = número de botones (índice 2 dentro de los parámetros)
    assert '2' in args[2], \
        f"Se esperaban 2 botones, se encontró: {args[2]}"


# ── Pines ─────────────────────────────────────────────────────────────────────

def test_pin_subir_es_6():
    assert re.search(r'PIN_SUBIR\s*=\s*6', source()), \
        "PIN_SUBIR debe estar asignado al pin 6"


def test_pin_bajar_es_3():
    assert re.search(r'PIN_BAJAR\s*=\s*3', source()), \
        "PIN_BAJAR debe estar asignado al pin 3"


# ── Configuración de pines ────────────────────────────────────────────────────

def test_usa_input_pullup_en_subir():
    assert re.search(r'pinMode\s*\(\s*PIN_SUBIR\s*,\s*INPUT_PULLUP\s*\)', source()), \
        "PIN_SUBIR debe configurarse como INPUT_PULLUP (común a GND)"


def test_usa_input_pullup_en_bajar():
    assert re.search(r'pinMode\s*\(\s*PIN_BAJAR\s*,\s*INPUT_PULLUP\s*\)', source()), \
        "PIN_BAJAR debe configurarse como INPUT_PULLUP (común a GND)"


# ── Lógica de lectura ─────────────────────────────────────────────────────────

def test_logica_activo_bajo():
    """Con pull-up + común a GND, el botón activo es LOW → negación con '!'."""
    assert '!digitalRead(PIN_SUBIR)' in source() or \
           '! digitalRead(PIN_SUBIR)' in source(), \
        "La lectura de PIN_SUBIR debe invertirse (activo bajo con INPUT_PULLUP)"
    assert '!digitalRead(PIN_BAJAR)' in source() or \
           '! digitalRead(PIN_BAJAR)' in source(), \
        "La lectura de PIN_BAJAR debe invertirse (activo bajo con INPUT_PULLUP)"


# ── Optimización USB ──────────────────────────────────────────────────────────

def test_send_state_condicional():
    """El reporte HID solo se envía cuando hay un cambio de estado."""
    src = source()
    assert 'estadoAnterior' in src, \
        "Debe existir un array/variable estadoAnterior para comparar cambios"
    assert 'Joystick.sendState()' in src, \
        "Debe llamar a Joystick.sendState() para enviar el reporte HID"


# ── Botones HID correctos ─────────────────────────────────────────────────────

def test_boton_0_es_subir():
    assert re.search(r'setButton\s*\(\s*0\s*,\s*subir\s*\)', source()), \
        "El botón HID 0 debe mapearse a 'subir'"


def test_boton_1_es_bajar():
    assert re.search(r'setButton\s*\(\s*1\s*,\s*bajar\s*\)', source()), \
        "El botón HID 1 debe mapearse a 'bajar'"


# ── Polling ───────────────────────────────────────────────────────────────────

def test_tiene_delay():
    assert re.search(r'delay\s*\(\s*10\s*\)', source()), \
        "Debe tener delay(10) para polling a ~100 Hz"

File: tests/test_palanca_secuencial.mjs
/**
 * Tests para palanca_secuencial.ino
 * Ejecutar con: node --test tests/test_palanca_secuencial.mjs
 *
 * Verifica que el código cumple la especificación de hardware:
 *   - Pin 6 = Subir (botón HID 0)
 *   - Pin 3 = Bajar  (botón HID 1)
 *   - 2 botones HID
 *   - INPUT_PULLUP (común a GND → activo bajo)
 *   - Envío condicional de estado HID
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const INO_PATH = join(ROOT, 'palanca_secuencial', 'palanca_secuencial.ino');

function src() {
  return readFileSync(INO_PATH, 'utf-8');
}

// ── Existencia del archivo ────────────────────────────────────────────────────

test('El archivo palanca_secuencial.ino existe', () => {
  assert.ok(existsSync(INO_PATH), `No se encontró ${INO_PATH}`);
});

// ── Dependencia HID ───────────────────────────────────────────────────────────

test('Incluye ArduinoJoystickLibrary', () => {
  assert.ok(src().includes('#include <Joystick.h>'),
    'Debe incluir #include <Joystick.h>');
});

// ── Configuración del joystick ────────────────────────────────────────────────

test('Joystick declara 2 botones', () => {
  // El constructor es: Joystick_ Joystick( ... )
  const match = src().match(/Joystick_\s+\w+\s*\(([^;]+?)\);/s);
  assert.ok(match, 'No se encontró el constructor Joystick_()');
  const args = match[1].split(',').map(s => s.trim());
  // 3er argumento (índice 2) = número de botones
  assert.ok(args[2].includes('2'), `Se esperaban 2 botones, se encontró: ${args[2]}`);
});

// ── Pines ─────────────────────────────────────────────────────────────────────

test('PIN_SUBIR está asignado al pin 6', () => {
  assert.ok(/PIN_SUBIR\s*=\s*6/.test(src()),
    'PIN_SUBIR debe ser el pin 6');
});

test('PIN_BAJAR está asignado al pin 3', () => {
  assert.ok(/PIN_BAJAR\s*=\s*3/.test(src()),
    'PIN_BAJAR debe ser el pin 3');
});

// ── INPUT_PULLUP ──────────────────────────────────────────────────────────────

test('PIN_SUBIR configurado como INPUT_PULLUP', () => {
  assert.ok(/pinMode\s*\(\s*PIN_SUBIR\s*,\s*INPUT_PULLUP\s*\)/.test(src()),
    'PIN_SUBIR debe usar INPUT_PULLUP (común a GND)');
});

test('PIN_BAJAR configurado como INPUT_PULLUP', () => {
  assert.ok(/pinMode\s*\(\s*PIN_BAJAR\s*,\s*INPUT_PULLUP\s*\)/.test(src()),
    'PIN_BAJAR debe usar INPUT_PULLUP (común a GND)');
});

// ── Lógica activo bajo ────────────────────────────────────────────────────────

test('Lectura de PIN_SUBIR es activo bajo (negación)', () => {
  assert.ok(src().includes('!digitalRead(PIN_SUBIR)'),
    'Con INPUT_PULLUP + GND debe invertirse: !digitalRead(PIN_SUBIR)');
});

test('Lectura de PIN_BAJAR es activo bajo (negación)', () => {
  assert.ok(src().includes('!digitalRead(PIN_BAJAR)'),
    'Con INPUT_PULLUP + GND debe invertirse: !digitalRead(PIN_BAJAR)');
});

// ── Mapeo de botones HID ──────────────────────────────────────────────────────

test('Botón HID 0 mapeado a subir', () => {
  assert.ok(/setButton\s*\(\s*0\s*,\s*subir\s*\)/.test(src()),
    'setButton(0, subir) debe mapear el botón de subir');
});

test('Botón HID 1 mapeado a bajar', () => {
  assert.ok(/setButton\s*\(\s*1\s*,\s*bajar\s*\)/.test(src()),
    'setButton(1, bajar) debe mapear el botón de bajar');
});

// ── Optimización USB ──────────────────────────────────────────────────────────

test('Usa estadoAnterior para comparar cambios', () => {
  assert.ok(src().includes('estadoAnterior'),
    'Debe existir estadoAnterior[] para envío condicional');
});

test('Llama a Joystick.sendState()', () => {
  assert.ok(src().includes('Joystick.sendState()'),
    'Debe enviar el reporte HID con sendState()');
});

// ── Polling ───────────────────────────────────────────────────────────────────

test('Tiene delay(10) para polling a ~100 Hz', () => {
  assert.ok(/delay\s*\(\s*10\s*\)/.test(src()),
    'Debe tener delay(10)');
});
